### DISCUSSION TITLE
What about upgrading?

### DISCUSSION THREAD

**Discussion by smulikHakipod** [berops/claudie#1011] [Q&A] [answered]
You guys make great work! Thanks a lot!
One of the most difficult issues I encountered with K8S in general is upgrading, upgrading both K8S version and OS packages and components. How should I approach upgrading a Claudie cluster? 
Thanks!

**Comment by MarioUhrik** [comment]
Hi @smulikHakipod, we've just opened an issue to add this documentation into Claudie https://github.com/berops/claudie/issues/1012, after which we'll officially commit to supporting a workflow that will be present in the future doc.
As of now, we have no officially supported workflow for this, but we know there are a handful of unofficial ways that we've personally tried, and that will probably be the basis of how the future doc will be written.

K8s version in Claudie is configured via the `version` field here:
https://github.com/berops/claudie/blob/master/docs/input-manifest/api-reference.md#cluster-k8s

Machine image (and therefore OS version and packages) is configured via the `image` field here:
https://github.com/berops/claudie/blob/master/docs/input-manifest/api-reference.md#dynamic

Obviously, you could delete a cluster and then remake it on an incremented version, or with changed machine image, but that's not what you're asking.

In order to do an upgrade without recreating the cluster:
- Simply incrementing the `version` field of the InputManifest will cause Claudie component KubeEleven (KubeOne, under the hood) to trigger a workflow that upgrades the cluster. KubeOne is fairly good at this, so your cluster will quite safely upgrade to the newly specified version. That's all there is to it. Still, I recommend being smart/careful about this, e.g. only increment one version at a time, etc. Also, make sure that the version you want to switch to is supported, as documented here https://github.com/berops/claudie/blob/master/docs/input-manifest/api-reference.md#cluster-k8s
- Changing the `image` field of the InputManifest will cause Claudie to start replacing the old nodes by new nodes. At the moment, as far as I know, there isn't any sort of a gentle "rollout" of this change, so it might cause some unavailability. For example, it's theoretically possible it might shut down all 3 out of 3 workers within a node pool all at the same time. If such a :cowboy_hat_face: approach is acceptable, then this is the simplest way. If you'd like to do this in a more careful manner, you can add in some intermediate steps into the process to ensure HA during the upgrade operation - for example, add a new nodepool with upgraded machine image, ensure it works, and then delete the old nodepool - all via the InputManifest.

In short, updating the InputManifest generally reflects the change into the infrastructure, but in some cases not in the most gentle way possible.

**Comment by jakubhlavacka** [comment, marked as answer]
The updating process is already described in our [documentation](https://docs.claudie.io/latest/update/update/).
